### PR TITLE
Individual esa fix

### PR DESCRIPTION
**Pull request by kathrynkodama**
Added install.individual.esa to map based install kernel to allow for an individual esa file to be specified in a pom.xml

### CODE REVIEW

File: dev/com.ibm.ws.install/src/com/ibm/ws/install/internal/InstallKernelMap.java [hunk 1]
```diff
@@ -520,6 +551,41 @@ private static boolean containsIgnoreCase(Collection<String> featureToInstall, S
...
+                    }
+
+                }
+            } catch (NullPointerException e) {
+                data.put(ACTION_RESULT, ERROR);
+                data.put(ACTION_ERROR_MESSAGE, e.getMessage());
+                data.put(ACTION_EXCEPTION_STACKTRACE, ExceptionUtils.stacktraceToString(e));
+            }
+
+            for (String feature : featureToInstall) {
```
Comment (ericglau) [inline dev/com.ibm.ws.install/src/com/ibm/ws/install/internal/InstallKernelMap.java]: Remove println

File: dev/com.ibm.ws.install/src/com/ibm/ws/install/internal/InstallKernelMap.java [hunk 2]
```diff
@@ -708,4 +774,70 @@ private Integer uninstall() {
...
+            while (zipEntries.hasMoreElements()) {
+                ZipEntry nextEntry = zipEntries.nextElement();
+                if ("OSGI-INF/SUBSYSTEM.MF".equalsIgnoreCase(nextEntry.getName())) {
+                    subsystemEntry = nextEntry;
+                    break;
+                }
+            }
+            if (subsystemEntry == null) {
+                ;
+            } else {
```
Comment (ericglau) [inline dev/com.ibm.ws.install/src/com/ibm/ws/install/internal/InstallKernelMap.java]: Just use `if (subsystemEntry != null) { Manifest m ...... }`

File: dev/com.ibm.ws.install/src/com/ibm/ws/install/internal/InstallKernelMap.java [hunk 3]
```diff
@@ -708,4 +774,70 @@ private Integer uninstall() {
         }
         return OK;
     }
-}
+
+    private static void getShortNameFromManifest(File esa, Map<String, String> shortNameMap) throws IOException {
```
Comment (ericglau) [inline dev/com.ibm.ws.install/src/com/ibm/ws/install/internal/InstallKernelMap.java]: Change method name to be more descriptive, since it is currently named "get..." but does not return anything.  Perhaps call it "populateShortNameFromManifest"

File: dev/com.ibm.ws.install/src/com/ibm/ws/install/internal/InstallKernelMap.java [hunk 4]
```diff
@@ -708,4 +774,70 @@ private Integer uninstall() {
...
+            }
+        } finally {
+            if (zip != null) {
+                zip.close();
+            }
+        }
+    }
+
+    //accept List of Files
+    private File individualESAInstall(Path generatedJson, Map<String, String> shortNameMap) throws IOException, BuildException, RepositoryException {
```
Comment (ericglau) [inline dev/com.ibm.ws.install/src/com/ibm/ws/install/internal/InstallKernelMap.java]: Change method name to be more descriptive, since the current name doesn't tell you what it does.  It doesn't install anything at this point.

File: dev/com.ibm.ws.install/src/com/ibm/ws/install/internal/InstallKernelMap.java [hunk 5]
```diff
@@ -708,4 +774,70 @@ private Integer uninstall() {
...
+            getShortNameFromManifest(esa, shortNameMap);
+
+            SingleFileRepositoryConnection mySingleFileRepo = null;
+            if (singleJson.exists()) {
+                mySingleFileRepo = new SingleFileRepositoryConnection(singleJson);
+            } else {
+                try {
+                    mySingleFileRepo = SingleFileRepositoryConnection.createEmptyRepository(singleJson);
+                } catch (IOException e) {
+                    e.printStackTrace();
```
Comment (ericglau) [inline dev/com.ibm.ws.install/src/com/ibm/ws/install/internal/InstallKernelMap.java]: Don't print stacktrace, should handle the exception

File: dev/com.ibm.ws.install/src/com/ibm/ws/install/internal/InstallKernelMap.java [hunk 6]
```diff
@@ -708,4 +774,70 @@ private Integer uninstall() {
...
+                }
+            }
+            System.out.println("Esa name: " + esa.getName());
```
Comment (ericglau) [inline dev/com.ibm.ws.install/src/com/ibm/ws/install/internal/InstallKernelMap.java]: Remove println

File: dev/com.ibm.ws.install/src/com/ibm/ws/install/internal/InstallKernelMap.java [hunk 7]
```diff
@@ -520,6 +551,41 @@ private static boolean containsIgnoreCase(Collection<String> featureToInstall, S
...
+                if (data.get(INSTALL_INDIVIDUAL_ESAS).equals(Boolean.TRUE)) {
+                    Path tempDir = Files.createTempDirectory("generatedJson");
+                    tempDir.toFile().deleteOnExit();
+                    Map<String, String> shortNameMap = new HashMap<String, String>();
+                    File individualEsaJson = individualESAInstall(tempDir, shortNameMap);
+                    RepositoryConnection repo = new SingleFileRepositoryConnection(individualEsaJson);
+                    repoList.add(repo);
+                    for (String feature : featureToInstall) {
+                        if (feature.endsWith(".esa")) {
+                            if (shortNameMap.containsKey(feature)) {
```
Comment (ericglau) [inline dev/com.ibm.ws.install/src/com/ibm/ws/install/internal/InstallKernelMap.java]: The `shortNameMap` was populated by `individualESAInstall()` with the keys being absolute paths to the ESA files.  But can ESA files be specified (e.g. in the Maven plugin's `<features>` section) with relative paths?  (You can try this with the existing plugin, which will use installUtility for individual esas.)  If so, then the `feature` string here could be a relative path, so `shortNameMap.containsKey(feature)` would be false even though we want it to match.  You could get around this by comparing canonical paths instead (not absolute paths, because absolute paths would not resolve things such as "." or "..")

Comment (kathrynkodama) [reply dev/com.ibm.ws.install/src/com/ibm/ws/install/internal/InstallKernelMap.java]: This was tested and ESA files cannot be specified with relative paths with installUtility.

Comment (ericglau) [reply dev/com.ibm.ws.install/src/com/ibm/ws/install/internal/InstallKernelMap.java]: What if they have absolute paths but with things such as "." or "..".  For example, `/folder/../folder/file.esa` and `/folder/file.esa` point to the same file, but comparing absolute paths would give false.  So we should still use canonical paths (so that both will be resolved to `/folder/file.esa`)

Comment (kathrynkodama) [reply dev/com.ibm.ws.install/src/com/ibm/ws/install/internal/InstallKernelMap.java]: This structure also results in a failed to invoke error.

File: dev/com.ibm.ws.install/src/com/ibm/ws/install/internal/InstallKernelMap.java (hunk 2 above)
Comment (kathrynkodama) [reply dev/com.ibm.ws.install/src/com/ibm/ws/install/internal/InstallKernelMap.java]: This change results in error: Only one "/" symbol is allowed in the resourceString, do you have alternate suggestions? Individual esa install works when this change is reverted.

File: dev/com.ibm.ws.install/src/com/ibm/ws/install/internal/InstallKernelMap.java [hunk 8]
```diff
@@ -708,4 +764,68 @@ private Integer uninstall() {
...
+                shortNameMap.put(esa.getAbsolutePath(), shortNameAttr);
+            }
+        } finally {
+            if (zip != null) {
+                zip.close();
+            }
+        }
+    }
+
+    //accept List of Files
```
Comment (ericglau) [inline dev/com.ibm.ws.install/src/com/ibm/ws/install/internal/InstallKernelMap.java]: Add javadoc style comments.  Generally they are not needed for private methods, but it would help with readability in this case since the method does quite a few things.

File: dev/com.ibm.ws.install/src/com/ibm/ws/install/internal/InstallKernelMap.java [hunk 9]
```diff
@@ -708,4 +764,68 @@ private Integer uninstall() {
...
+    private File individualESAInstall(Path generatedJson, Map<String, String> shortNameMap) throws IOException, BuildException, RepositoryException, InstallException {
```
Comment (ericglau) [inline dev/com.ibm.ws.install/src/com/ibm/ws/install/internal/InstallKernelMap.java]: Perhaps rename method to something like generateJsonsFromIndividualESAs

Comment (ericglau) [inline dev/com.ibm.ws.install/src/com/ibm/ws/install/internal/InstallKernelMap.java]: Rename parameter `generatedJson` since the parameter is not a json but is just the temp dir where the jsons will be stored.  Perhaps use `jsonDirectory`

File: dev/com.ibm.ws.install/resources/com/ibm/ws/install/internal/resources/InstallKernel.nlsprops [hunk 10]
```diff
@@ -771,6 +771,14 @@ ERROR_UNKNOWN_OSLIST_VALUE=CWWKF1387E: The value {0} is not a valid entry for th
 ERROR_UNKNOWN_OSLIST_VALUE.explanation=This value will be ignored as it is not a recognized operating system name.
 ERROR_UNKNOWN_OSLIST_VALUE.useraction=Please modify or remove this value from the feature manifest file.
 
+ERROR_ESA_NOT_FOUND=CWWKF1388E: The specified ESA file did not contain any features.
+ERROR_ESA_NOT_FOUND.explanation=A feature will not be installed from the ESA file.
+ERROR_ESA_NOT_FOUND.useraction=Ensure that the individual ESA file provided is valid and contains a valid feature.
```
Comment (ericglau) [inline dev/com.ibm.ws.install/resources/com/ibm/ws/install/internal/resources/InstallKernel.nlsprops]: Change to:
```
ERROR_ESA_NOT_FOUND.useraction=Ensure that the ESA file provided is valid and contains a feature.
```

File: dev/com.ibm.ws.install/resources/com/ibm/ws/install/internal/resources/InstallKernel.nlsprops [hunk 11]
```diff
@@ -771,6 +771,14 @@ ERROR_UNKNOWN_OSLIST_VALUE=CWWKF1387E: The value {0} is not a valid entry for th
 ERROR_UNKNOWN_OSLIST_VALUE.explanation=This value will be ignored as it is not a recognized operating system name.
 ERROR_UNKNOWN_OSLIST_VALUE.useraction=Please modify or remove this value from the feature manifest file.
 
+ERROR_ESA_NOT_FOUND=CWWKF1388E: The specified ESA file did not contain any features.
```
Comment (ericglau) [inline dev/com.ibm.ws.install/resources/com/ibm/ws/install/internal/resources/InstallKernel.nlsprops]: It would be helpful for the message to have a parameter to say what file it is referring to.
e.g. `The specified ESA file {0} did not contain any features.` where {0} is the file path

File: dev/com.ibm.ws.install/resources/com/ibm/ws/install/internal/resources/InstallKernel.nlsprops [hunk 12]
```diff
@@ -771,6 +771,14 @@ ERROR_UNKNOWN_OSLIST_VALUE=CWWKF1387E: The value {0} is not a valid entry for th
...
+ERROR_ESA_NOT_FOUND.explanation=A feature will not be installed from the ESA file.
+ERROR_ESA_NOT_FOUND.useraction=Ensure that the individual ESA file provided is valid and contains a valid feature.
+
+ERROR_SINGLE_REPO_CONNECTION_FAILED=CWWKF1389E: Unable to establish a single file repository connection for the single JSON.
```
Comment (ericglau) [inline dev/com.ibm.ws.install/resources/com/ibm/ws/install/internal/resources/InstallKernel.nlsprops]: [SingleFileRepositoryConnection](https://github.com/OpenLiberty/open-liberty/blob/integration/dev/com.ibm.ws.repository/src/com/ibm/ws/repository/connections/SingleFileRepositoryConnection.java) throws IOException if the json already exists or there is a problem creating the file.  We can assume it doesn't already exist since InstallKernelMap line 810 checked for that, so the issue would be a problem creating the file.  So the message can be like:

```
ERROR_SINGLE_REPO_CONNECTION_FAILED=CWWKF1389E: Unable to create a JSON file at {0} for the ESA file {1}.
ERROR_SINGLE_REPO_CONNECTION_FAILED.explanation=The feature in the ESA file will not be installed.
ERROR_SINGLE_REPO_CONNECTION_FAILED.useraction=Ensure that write access is available for the directory where the JSON file will be created.
```

File: dev/com.ibm.ws.install/src/com/ibm/ws/install/internal/InstallKernelMap.java [hunk 13]
```diff
@@ -708,4 +764,79 @@ private Integer uninstall() {
...
+
+        String dir = jsonDirectory.toString();
+        List<File> esas = (List<File>) data.get(INDIVIDUAL_ESAS);
+        File singleJson = new File(dir + "/SingleJson.json");
+
+        for (File esa : esas) {
+            try {
+                populateShortNameFromManifest(esa, shortNameMap);
+            } catch (IOException e) {
+                throw new InstallException(Messages.INSTALL_KERNEL_MESSAGES.getLogMessage("ERROR_ESA_NOT_FOUND", esa.getCanonicalPath()));
```
Comment (ericglau) [inline dev/com.ibm.ws.install/src/com/ibm/ws/install/internal/InstallKernelMap.java]: Use `esa.getAbsolutePath()` for the message, otherwise getting canonical path could throw another exception which is not caught.

File: dev/com.ibm.ws.install/src/com/ibm/ws/install/internal/InstallKernelMap.java [hunk 14]
```diff
@@ -708,4 +764,79 @@ private Integer uninstall() {
...
+
+            SingleFileRepositoryConnection mySingleFileRepo = null;
+            if (singleJson.exists()) {
+                mySingleFileRepo = new SingleFileRepositoryConnection(singleJson);
+            } else {
+                try {
+                    mySingleFileRepo = SingleFileRepositoryConnection.createEmptyRepository(singleJson);
+                } catch (IOException e) {
+                    throw new InstallException(Messages.INSTALL_KERNEL_MESSAGES.getLogMessage("ERROR_SINGLE_REPO_CONNECTION_FAILED", dir,
+                                                                                              esa.getCanonicalPath()));
```
Comment (ericglau) [inline dev/com.ibm.ws.install/src/com/ibm/ws/install/internal/InstallKernelMap.java]: Use `esa.getAbsolutePath()` for the message, otherwise getting canonical path could throw another exception which is not caught.

File: dev/com.ibm.ws.app.manager.springboot/src/com/ibm/ws/app/manager/springboot/internal/LibIndexCache.java [hunk 15]
```diff
@@ -66,7 +66,7 @@ protected void setAdaptableModuleFactory(AdaptableModuleFactory adaptableFactory
         this.adaptableFactory = adaptableFactory;
     }
 
-    public File getLibrary(Map.Entry<String, String> LibIndexEntry) {
+    private File getLibrary(Map.Entry<String, String> LibIndexEntry) {
```
Comment (ericglau) [inline dev/com.ibm.ws.app.manager.springboot/src/com/ibm/ws/app/manager/springboot/internal/LibIndexCache.java]: This and some other files seem unrelated to your changes.  Can you merge or rebase from integration?

**Comment by kathrynkodama** [comment]
build: https://wasrtc.hursley.ibm.com:9443/jazz/web/projects/WS-CD#action=com.ibm.team.build.viewResult&id=_iPJOkMM-Eei9re3pB293Wg&tab=com.ibm.team.build.web.ui.internal.editors.result.overview.OverviewPage
